flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java

```diff
@@ -1061,6 +1063,20 @@ void markFinished(Map<String, Accumulator<?, ?>> userAccumulators, IOMetrics met
 							IntermediateResultPartition[] allPartitions = finishedPartition
 									.getIntermediateResult().getPartitions();
 
+							PartitionTable<ResourceID> partitionTable = vertex.getExecutionGraph().getPartitionTable();
+							for (IntermediateResultPartition partition : allPartitions) {
+								ResultPartitionDeploymentDescriptor descriptor = producedPartitions.get(partition.getPartitionId());
+								if (!descriptor.isReleasedOnConsumption()) {
```

Comment:
At the end we should also take into account whether this partition actually occupies local resources (see https://issues.apache.org/jira/browse/FLINK-12890) otherwise TM and its connection can be released independently.